### ISSUE TITLE
[Suggestion] Servers folder plugin

### ISSUE THREAD

**Issue by cyyynthia** [powercord-org/powercord#34]

**Comment by devoxin** [comment]
fuck yes please i will give you all my money just to implement this i need this

**Comment by melmsie** [comment]
I will not give you all my money, but I'll give you 5.3% of my total love

**Comment by GeoffreyWesthoff** [comment]
Bonus points if it looks like 
![image](https://user-images.githubusercontent.com/22159739/50386279-f0466980-06e3-11e9-929a-f9c193eda30e.png)

**Comment by melmsie** [comment]
bonus bonus points if it has folders AND search bar

**Comment by stayberri** [comment]
if this becomes a thing i shall donate 0.05% of my total affection (beat that melms)

**Comment by aetheryx** [comment]
search bar: `Ctrl`+`k`

**Comment by melmsie** [comment]
>keyboard shortcuts
what is this, i3

**Comment by GeoffreyWesthoff** [comment]
I was thinking about the collections/folders not the search bar

**Comment by Stylite** [comment]
stackable groups or folders would be hOT

**Comment by GeoffreyWesthoff** [comment]
Porque no los dos? @Stylite

**Comment by cyyynthia** [comment]
I'll do my best to implement this (or someone else if someone wants) as best as possible
I'll also include hidden servers (for emoji-only servers) and *maybe* subfolders for ultra sorting uwu
Tho I can't work on it until wednesday and it'll be included with the new Settings API and plugin system

**Comment by melmsie** [comment]
stack-able groups wouldn't be much better if someone has lots of servers, folders or bust

**Comment by cyyynthia** [comment]
And ofc styles of folders will be selectable in settings tab, and hidden servers will be also managed there uwu

**Comment by GeoffreyWesthoff** [comment]
@Bowser65 Would it be truly necessary to have hidden servers? A folder at the bottom would suffice for a lot of people I reckon

**Comment by cyyynthia** [comment]
As it should not be hard to implement that, why not implement it and let people decide if they want to use it or not /shrug

**Comment by devoxin** [comment]
I personally like folders and collections. Although collections could look a bit cluttered, especially with lots of servers.

Dividers doesn't solve the issue 100% as there would probably still be a bit of scrolling involved, and would still look fairly cluttered if you have lots of servers.

Sliders could be an issue if you're trying to get to the far-right of the list, as your mouse doesn't have a lot of room to move.

Stacked, as per what Melmsie said, wouldn't be too great for those with lots of servers. I.e. it would look cluttered and maybe be hard to distinguish servers in the stack if they mix with your other stacks/servers.

Quick search is nice, although doesn't solve the issue of being able to group servers into categories, i.e. emote servers, twitch servers etc.

My favorite idea is a mix between sliders, folders, and search bar.

Search bar would allow you to quickly find a server rather than navigating through multiple folders, which would be great for servers that are accessed often. Not too good for those whose name changes often (think: emote servers).

Folders would be great to keep the list looking clean(er), while potentially having indicators for easy identification of what the folder contains (such as emote servers etc. Perhaps allow users to set image icons for the folders)

Sliders could come into play when a user selects a folder. A panel extends horizontally across maybe half of the screen, which shows all the servers in that folder. A user can then use the scroll-wheel to navigate through the list, or perhaps drag-and-slide with the cursor (this could present an issue when trying to order servers, though). The extended panel would help here, offering more room to slide the cursor across.

Alternatively, a vertical panel could extend to the right of the server list, that mimics the server list exactly, except it'd show your servers, and you'd be able to scroll through them as normal.

Server ordering would be an additional bonus, although I expect that *any* of the above mentioned in this post, and others, would take time to be implement (if not outright impossible at that, lol).

**Comment by cyyynthia** [comment]
tldr

**Comment by cyyynthia** [comment]
Well this feature will probably require a lot of work due to the amount of parameters to include (mentions, unreads, icons, settings, etc etc)

I'll do my best to make this as best as possible, since it's a feature that generates lots of hype haha

Thanks for your comments and ideas, I'll keep this thread updated with progress I guess :heart:

**Comment by cyyynthia** [comment]
Updates:

We tried to inject into `Guilds` component, and we quickly figured out that injecting a new component is as always a pretty easy task, but DnD library Discord uses is way harder to inject into. It produces weird behavior and will trigger the API on every sorting updates. Just patching would require lots of injection points and would start being ugly/heavy

Instead, we plan to completely re-create `Guilds` component, with our own DnD lib and our own behavior regarding server sorting updates on Discord API. That will require to at least reproduce DOM structure (with only .pc- classes) to make sure themes will still work (I plan to transform `@powercordCompile` in `@powercordIgnore` for themes). For the end user he'll not really notice anything if we do stuff properly, but will be way easier for us to make this feature possible.

With this we can also implement an API endpoint on Powercord side to save your Guild sorting preferences, so it'll be synced to any PC if you linked your PC account to Discord. For non PC clients we'll do a basic but coherent logic for updating sorting preferences: `instead of folders, servers in the orders they are in the folder`

Coming soon:tm:

**Comment by cyyynthia** [comment]
Development officially started here: https://github.com/aetheryx/powercord/tree/experimental-servers-folder
For now it just patches Discord sidebar guild list and replaces it by the Powercord's one, without drag n drop for now but soon:tm:

**Comment by cyyynthia** [comment]
![yes](https://cdn.discordapp.com/attachments/517348857219383308/533411353440944128/Ahhhhhhhhhhhhhhhhh.gif)

**Comment by cyyynthia** [comment]
Updates:

Guild hiding is now live! Right click on a guild to make it disappear. A small component at the top will allow you to make them visible, or know how many hidden pings is there.

Internally, Powercord uses a fully custom Guild component, with custom drag & drop to reproduce Discord's behaviour. Folders are not a thing yet, but soon:tm:

**Comment by cyyynthia** [comment]
Note: it does not break with new "Server Discovery" feature, so don't worry about that
